libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift

```diff
@@ -237,35 +282,396 @@ extension SFRestAPI {
                 resolver.fulfill(self.api!.request(forQueryAll: soql))
             }
         }
-    }
-    
-    /**
-     Send api wrapped in a promise.
-     
-     ```
-     let restApi = SFRestAPI.sharedInstance()
-     restApi.Factory.describe(objectType: "Account")
-     .then { request in
+        
+        /**
+         A factory method for sosl object request.
+         ```
+          restApi.Promises.search(sosl: sosl)
+         .then { (request) in
+            ...
+         }
+         ```
+         - Returns:  SFRestRequest wrapped in a promise.
+         */
+        public func search(sosl: String) -> Promise<SFRestRequest> {
+            return  Promise(.pending) {  resolver in
+                resolver.fulfill(self.api!.request(forSearch: sosl))
+            }
+        }
+        
+        /**
+         A factory method for Search Scope And Order request.
+         ```
+         restApi.Promises.searchScopeAndOrder()
+         .then { (request) in
+            ...
+         }
+         ```
+         - Returns:  SFRestRequest wrapped in a promise.
+         */
+        public func searchScopeAndOrder() -> Promise<SFRestRequest> {
+            return  Promise(.pending) {  resolver in
+                resolver.fulfill(self.api!.requestForSearchScopeAndOrder())
+            }
+        }
+        
+        /**
+         A factory method for Search Result Layout request.
+         ```
+         restApi.Promises.searchResultLayout(objectList: "Account","Contact")
+         .then { (request) in
+            ...
+         }
+         ```
+         - Returns:  SFRestRequest wrapped in a promise.
+         */
+        public func searchResultLayout(objectList: String...) -> Promise<SFRestRequest> {
+            return self.searchResultLayout(objectList: objectList)
+        }
+        
+        /**
+         A factory method for Search Result Layout request.
+         ```
+         restApi.Promises.searchResultLayout(objectList: ["Account","Contact"])
+         .then { (request) in
+         ...
+         }
+         ```
+         - Returns:  SFRestRequest wrapped in a promise.
+         */
+        public func searchResultLayout(objectList: [String]) -> Promise<SFRestRequest> {
+            return  Promise(.pending) {  resolver in
+                resolver.fulfill(
+                    self.api!.request(forSearchResultLayout: objectList.joined(separator: ",")))
+            }
+        }
+        
+        /**
+         A factory method for Batch request.
+         ```
+         restApi.Promises.batch(requests: request1,request2, haltOnError: Bool = false)
+         .then { (request) in
+         ...
+         }
+         ```
+         - Returns:  SFRestRequest wrapped in a promise.
+         */
+        public func batch(requests: SFRestRequest..., haltOnError: Bool = false) -> Promise<SFRestRequest> {
```

Comment:
Interesting - the ... arg does not have to be last?

Reply:
@wmathurin it doesn't have to be.